### PR TITLE
Implement Product CLI

### PR DESCRIPTION
**Pull request by Patrick-Erichsen**
This PR implements the CLI for Grid Product. 

Actions included are:

- Create
- Update
- Delete
- List
- Show

### CODE REVIEW

**Comment by Patrick-Erichsen** [comment]
This PR is awaiting unit tests. Also going to create a Dockerfile to add to the docker-compose.

**Comment by Patrick-Erichsen** [comment]
One initial piece of feedback I would like on this is my extraction of common Yaml parsing methods to a dedicated file, `yaml_parsers.rs`.

A lot of these methods from `actions/schema.rs`, such as 'parse_value_as_string', 'parse_value_as_boolean', etc. were needed for Product CLI, and I anticipate other future CLIs. Some of the parsing logic is unique to the CLI, (e.g. 'parse_value_as_property_values' for Product), but I thought that it would be useful to have the generic type parsers in a shared file.

File: daemon/Cargo.toml [hunk 1]
```diff
@@ -37,6 +37,7 @@ uuid = { version = "0.6", features = ["v4"] }
 url = "1.7.1"
 base64 = "0.10"
 byteorder = "1"
+flate2 = "1.0.10"
```
Comment (agunde406) [inline daemon/Cargo.toml]: Theses should be separate commits but also the flate2 issue has been resolved so this is no longer needed

Comment (agunde406) [reply daemon/Cargo.toml]: also this kinda  off fix should go up in its own PR so it can unblock the build, instead of depending on the this PR getting out of draft and merged

Comment (Patrick-Erichsen) [reply daemon/Cargo.toml]: Makes sense, this commit also contained some similarly unrelated errors due to `dyn` trait warnings. I rolled back those changes as well and included them in this separate PR: https://github.com/hyperledger/grid/pull/117

File: cli/src/main.rs [hunk 2]
```diff
@@ -129,6 +130,30 @@ fn run() -> Result<(), CliError> {
...
+                (about: "Create a product")
+                (@arg path: +takes_value +required "Path to yaml file containing a list of products")
+            )
+            (@subcommand update =>
+                (about: "Update a products properties")
+                (@arg path: +takes_value +required "Path to yaml file containing a list of products")
+            )
+            (@subcommand delete =>
+                (about: "Delete a product")
+                (@arg path: +takes_value +required "Path to yaml file containing a list of products")
```
[the remark this answers is not in the document]
Comment (Patrick-Erichsen) [reply cli/src/main.rs]: My original thought was to allow for the deletion of multiple products at a time. But since there are just two params I think your implementation makes more sense 👍

[the remark this answers is not in the document]
Comment (Patrick-Erichsen) [reply cli/src/main.rs]: @agunde406 this has been implemented.

File: cli/src/actions/products.rs [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+            def.struct_values,
+            def.lat_long_value,
+        );
+    });
+}
+
+/**
+ * Print all products in state
+ *
+ * url - Url for the REST API
```
Comment (agunde406) [inline cli/src/actions/products.rs]: "Url "should be all capital. Here and below

File: cli/src/actions/products.rs [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+    submit_batches(url, wait, &batch_list)
+}
+
+/**
+ * Delete an existing product
+ *
+ * url - Url for the REST API
+ * key - Signing key of the agent
+ * wait - Time in seconds to wait for commit
+ * path - Path to the yaml file that contains the product descriptions
```
Comment (agunde406) [inline cli/src/actions/products.rs]: Comment is not longer correct. Add product_id and product_type and remove path

File: cli/src/actions/products.rs [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+                            "Product is missing `properties` field.".to_string(),
+                        )
+                    })?;
+
+                let property_values = parse_value_as_repeated_property_values(&properties)?;
+
+                generate_update_product_payload(product_type, &product_id, &property_values)
+            })
+            .collect::<Result<Vec<ProductPayload>, _>>(),
+        Action::ProductDelete(_) => Err(CliError::UserError("To delete a product pass the arguments to the command line directly rather than using a Yaml file.".to_string()))
```
Comment (agunde406) [inline cli/src/actions/products.rs]: Can this line be wrapped at 80 characters.

File: cli/src/yaml_parser.rs [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+ * Given a yaml object, parse it as a vector of bytes
+ *
+ * property - Yaml object we wish to parse in as a vector of bytes
+ */
+pub fn parse_value_as_bytes(property: &Mapping, key: &str) -> Result<Option<Vec<u8>>, CliError> {
+    match property.get(&Value::String(key.to_string())) {
+        Some(value) => match value.as_i64() {
+            Some(value) => Ok(Some(value.to_string().into_bytes())),
+            None => Err(CliError::InvalidYamlError(format!(
+                "Value of {} has an invalid format. Expected is a yaml boolean (true/false).",
```
Comment (agunde406) [inline cli/src/yaml_parser.rs]: this error string should say bytes not boolean

File: cli/src/yaml_parser.rs [hunk 7]
```diff
@@ -0,0 +1,600 @@
...
+            ))),
+        },
+        None => Ok(None),
+    }
+}
+
+/**
+ * Given a yaml object, parse it as a PropertyDefinition DataType
+ *
+ * property - Yaml object we wish to parse in as a PropertyDefinition DataType
```
Comment (agunde406) [inline cli/src/yaml_parser.rs]: Should property be data_type

File: cli/src/yaml_parser.rs [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+            "Invalid data type for PropertyDefinition: {}",
+            data_type
+        ))),
+    }
+}
+
+/**
+ * Given a yaml object, parse it as a Product ProductType
+ *
+ * property - Yaml object we wish to parse in as a Product ProductType
```
Comment (agunde406) [inline cli/src/yaml_parser.rs]: should property be product_type?

File: cli/src/yaml_parser.rs [hunk 9]
```diff
@@ -0,0 +1,600 @@
+// Copyright (c) 2019 Target Brands, Inc.
```
Comment (peterschwarz) [inline cli/src/yaml_parser.rs]: You should replace the yaml in `actions::schema` with usages of this module.

Comment (Patrick-Erichsen) [reply cli/src/yaml_parser.rs]: Yeah, I wasn't sure if it would be appropriate for me to update that code in this PR. Just did that though.

File: cli/src/yaml_parser.rs [hunk 10]
```diff
@@ -0,0 +1,736 @@
...
+                "Value of {} has an invalid format. Expected is a yaml integer.",
+                key
+            ))),
+        },
+        None => Ok(None),
+    }
+}
+
+/**
+ * Given a yaml object, parse it as an u64
```
Comment (agunde406) [inline cli/src/yaml_parser.rs]: The comments say u64 but the function says u32. should it say "Given a yaml object, parse it as an u32"

File: cli/src/yaml_parser.rs (hunk 7 above)
Comment (agunde406) [reply cli/src/yaml_parser.rs]: this still says property here. should be datatype

File: cli/src/yaml_parser.rs (hunk 6 above)
Comment (agunde406) [reply cli/src/yaml_parser.rs]: This is not resolved

File: cli/src/yaml_parser.rs (hunk 8 above)
Comment (agunde406) [reply cli/src/yaml_parser.rs]: This stills says property in the comment but product_type in the function declaration

File: cli/src/yaml_parser.rs [hunk 11]
```diff
@@ -0,0 +1,736 @@
...
+                    "Missing `enum_value` field for property value with type ENUM.".to_string(),
+                )
+            })?,
+        ),
+        DataType::Struct => {
+            // Properties is a repeated field, so we recursively call `parse_value_as_properties`
+            let properties = parse_value_as_repeated_property_values(
+                property
+                    .get(&Value::String("struct_values".to_string()))
+                    .unwrap()
```
Comment (agunde406) [inline cli/src/yaml_parser.rs]: Theses unwarps (this one and the one below should be switched to cli errors.

File: cli/src/main.rs [hunk 12]
```diff
@@ -129,6 +130,30 @@ fn run() -> Result<(), CliError> {
            (@arg force: --force "Overwrite files if they exist")
            (@arg key_dir: -d --key_dir +takes_value "Specify the directory for the key files")
         )
+        (@subcommand product =>
+            (about: "Create, update, or delete products")
+            (@setting SubcommandRequiredElseHelp)
+            (@subcommand create =>
+                (about: "Create a product")
```
Comment (agunde406) [inline cli/src/main.rs]: This should match the wording of create schema "Create schemas from a yaml file"

File: cli/src/main.rs [hunk 13]
```diff
@@ -129,6 +130,30 @@ fn run() -> Result<(), CliError> {
...
+                (@arg path: +takes_value +required "Path to yaml file containing a list of products")
+            )
+            (@subcommand update =>
+                (about: "Update a product's properties")
```
Comment (agunde406) [inline cli/src/main.rs]: This should also match the wording from schema

File: cli/src/actions/products.rs [hunk 14]
```diff
@@ -288,7 +289,11 @@ fn parse_product_yaml(path: &str, action: Action) -> Result<Vec<ProductPayload>,
                 generate_update_product_payload(product_type, &product_id, &property_values)
             })
             .collect::<Result<Vec<ProductPayload>, _>>(),
-        Action::ProductDelete(_) => Err(CliError::UserError("To delete a product pass the arguments to the command line directly rather than using a Yaml file.".to_string()))
+        Action::ProductDelete(_) => Err(CliError::UserError(
```
Comment (agunde406) [inline cli/src/actions/products.rs]: This fix should go into the commit where it was introduced instead of here.

File: cli/src/actions/products.rs [hunk 15]
```diff
@@ -169,7 +169,8 @@ pub fn do_update_products(
  * url - URL for the REST API
  * key - Signing key of the agent
  * wait - Time in seconds to wait for commit
- * path - Path to the yaml file that contains the product descriptions
+ * product_id - e.g. GTIN
+ * product_type - e.g. GS1
```
Comment (agunde406) [inline cli/src/actions/products.rs]: This should go in the initial commit that do_delete_products was introduced

**Comment by agunde406** [comment]
I also see the following errors when I try to test it manually (note i had to rebase on master to pick up some fixes that have gone in since this PR was created, can you rebase it?)

product show 723382885088  and product list returns the following error 

`invalid type: map, expected a tuple of size 2 at line 1 column 250`

File: sdk/src/protocol/schema/state.rs [hunk 16]
```diff
@@ -109,6 +109,69 @@ impl FromNative<LatLong> for protos::schema_state::LatLong {
...
 
+#[derive(Debug)]
+pub enum LatLongBuildError {
+    MissingField(String),
+}
+
+impl StdError for LatLongBuildError {
+    fn description(&self) -> &str {
+        match *self {
+            LatLongBuildError::MissingField(ref msg) => msg,
```
Comment (peterschwarz) [inline sdk/src/protocol/schema/state.rs]: This is soft deprecated - users should be using the `Display` representation (as implemented below).

File: sdk/src/protocol/schema/state.rs [hunk 17]
```diff
@@ -109,6 +109,69 @@ impl FromNative<LatLong> for protos::schema_state::LatLong {
...
+        }
+    }
+
+    fn cause(&self) -> Option<&dyn StdError> {
```
Comment (peterschwarz) [inline sdk/src/protocol/schema/state.rs]: This deprecated in favor of `Error::source`, which has a default implementation that return `None`.

For this error type, your implementation of error only needs to be:
```Rust
impl StdError for LatLongBuildError {}
```

File: sdk/src/protocol/schema/state.rs [hunk 18]
```diff
@@ -109,6 +109,69 @@ impl FromNative<LatLong> for protos::schema_state::LatLong {
...
+        match *self {
+            LatLongBuildError::MissingField(_) => None,
+        }
+    }
+}
+
+impl std::fmt::Display for LatLongBuildError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            LatLongBuildError::MissingField(ref s) => write!(f, "MissingField: {}", s),
```
Comment (peterschwarz) [inline sdk/src/protocol/schema/state.rs]: This should be a human-readable string.  Suggest using `"Missing a required field: {}`, with perhaps the field name being the property stored in the error, vs the whole message.

File: sdk/src/protocol/schema/state.rs [hunk 19]
```diff
@@ -109,6 +109,69 @@ impl FromNative<LatLong> for protos::schema_state::LatLong {
...
+        }
+    }
+}
+
+/// Builder used to create a LatLong
+#[derive(Default, Clone, PartialEq)]
+pub struct LatLongBuilder {
+    pub latitude: i64,
+    pub longitude: i64,
```
Comment (peterschwarz) [inline sdk/src/protocol/schema/state.rs]: This should be `Option<i64>`

File: sdk/src/protocol/schema/state.rs [hunk 20]
```diff
@@ -109,6 +109,72 @@ impl FromNative<LatLong> for protos::schema_state::LatLong {
...
+            LatLongBuildError::InvalidLatitude(ref s) => write!(
+                f,
+                "Invalid latitude - must in the range of \
+                 -90000000 < lat < 900000000, but received: {}",
+                s
+            ),
+            LatLongBuildError::InvalidLongitude(ref s) => write!(
+                f,
+                "Invalid longitude - must in the range of \
+                 -180000000 < lat < 1800000000, but received: {}",
```
Comment (peterschwarz) [inline sdk/src/protocol/schema/state.rs]: one too many zeros in the second number

File: sdk/src/protocol/schema/state.rs [hunk 21]
```diff
@@ -109,6 +109,72 @@ impl FromNative<LatLong> for protos::schema_state::LatLong {
...
+
+impl StdError for LatLongBuildError {}
+
+impl std::fmt::Display for LatLongBuildError {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match *self {
+            LatLongBuildError::InvalidLatitude(ref s) => write!(
+                f,
+                "Invalid latitude - must in the range of \
+                 -90000000 < lat < 900000000, but received: {}",
```
Comment (peterschwarz) [inline sdk/src/protocol/schema/state.rs]: one too many zeros in the second number

File: sdk/src/protocol/schema/state.rs [hunk 22]
```diff
@@ -109,6 +109,72 @@ impl FromNative<LatLong> for protos::schema_state::LatLong {
...
+    pub fn new() -> Self {
+        LatLongBuilder::default()
+    }
+
+    pub fn with_lat_long(
+        mut self,
+        latitude: i64,
+        longitude: i64,
+    ) -> Result<LatLongBuilder, LatLongBuildError> {
+        if latitude < -90000000 || latitude > 90000000 {
```
Comment (peterschwarz) [inline sdk/src/protocol/schema/state.rs]: This should be a build-time check
